Time series logging

matthiasdiener opened the pull request:
Implements https://github.com/illinois-ceesd/planning/issues/57

matthiasdiener commented:
Here is how the output looks like (x-axis is simtime, y-axis is walltime per step):

![image](https://user-images.githubusercontent.com/1772435/92928379-9bd9a500-f404-11ea-918f-f9d6eaad6ff2.png)

matthiasdiener commented:
@inducer Is this going in the right direction (in particular examples/sod-mpi.py and mirgecom/steppers.py)?

matthiasdiener commented on mirgecom/steppers.py:
```diff
@@ -27,9 +27,47 @@
+        add_simulation_quantities, add_run_info, IntervalTimer,
+        set_dt, LogQuantity)
+
+class Pressure(LogQuantity):
```
Should we use a MultiLogQuantity here (for min,avg,max)?

inducer replied:
No, one at a time for now, as discussed.

inducer commented on mirgecom/steppers.py:
```diff
@@ -27,9 +27,47 @@
+
+  def __call__(self):
+    from functools import partial
+    _min = partial(self.discr.nodal_min, "vol")
```
Make sure that there's no MPI reduction in this.

matthiasdiener replied:
As far as I can see, there is none.

matthiasdiener commented on mirgecom/steppers.py:
```diff
@@ -27,9 +27,47 @@
+        set_dt, LogQuantity)
+
+class Pressure(LogQuantity):
+  def __init__(self, discr, current_state, eos):
```
I added this 

https://github.com/illinois-ceesd/mirgecom/pull/96/files#diff-181d5e42729a73c3035b94a08dc66983R91-R93

and

https://github.com/illinois-ceesd/mirgecom/pull/96/files#diff-181d5e42729a73c3035b94a08dc66983R32-R37

But I presume this is not what you had in mind?

matthiasdiener replied:
Fixed.

matthiasdiener commented:
This is ready for review

inducer commented on mirgecom/steppers.py:
```diff
@@ -27,9 +27,13 @@
 
 def advance_state(rhs, timestepper, checkpoint, get_timestep,
-                  state, t_final, t=0.0, istep=0):
+                  state, t_final, t=0.0, istep=0, logmgr=None, instrumenter=None):
```
I don't agree that integrating pyinstrument here is a good idea. Likely, you would want to profile over multiple time steps.

matthiasdiener replied:
I agree with your point, but from what I can see, pyinstrument is relatively inflexible, in that you can't break down the profiling results to a finer granularity after profiling.

matthiasdiener replied:
Removed pyinstrument.

inducer reviewed (commented):
Huh. Sorry. I had a review pending here that apparently didn't submit properly.

matthiasdiener commented:
> Huh. Sorry. I had a review pending here that apparently didn't submit properly.

Heh, I thought I had lost those comments :-(

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+                gd.quantity.set_state(state)
+
+
+class StateConsumer:
```
> * Possible solution: Have `__call__` reset `state` to `None`?

Done.

> * Also check the other stateful quantities in `logpyle`.

Not sure what you mean here.

inducer replied:
> Not sure what you mean here.

Please audit logpyle for potential uses of stale data.

matthiasdiener replied:
Per our discussion, I tested re-setting the time step to None and 0 in https://github.com/illinois-ceesd/logpyle/pull/32. This seems to work as expected, i.e. app terminates when `set_dt` is not called on next tick. Is this what you had in mind @inducer ?

inducer replied:
I'll worry about that.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self, state: ndarray):
+        self.state = state
```
Done.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+        cv = split_conserved(self.discr.dim, self.state)
+        dv = self.eos.dependent_vars(cv)
+
+        return self._myop(getattr(dv, self.quantity))
```
Added.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+        cv = split_conserved(self.discr.dim, self.state)
+        dv = self.eos.dependent_vars(cv)
+
+        return self._myop(getattr(dv, self.quantity))
```
Added a 'dim' parameter for these cases.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+
+        from mirgecom.euler import split_conserved
+
+        cv = split_conserved(self.discr.dim, self.state)
```
Would this be appropriate for a new PR (like, adding a ChemicalQuantity class)?

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,109 @@
+    """Logging support for physical quantities."""
+
+    def __init__(self, discr, eos, quantity, unit, op: str):
+        LogQuantity.__init__(self, f"{op}_{quantity}", unit)
```
I implemented a simple aggregator selection in https://github.com/illinois-ceesd/mirgecom/pull/96/commits/8040ac826035effc5722b275dddfdda610383562 Is this on the right track @inducer ?

matthiasdiener commented:
@inducer - I think you didn't submit your comments from yesterday regarding the version.sh script, but I think I addressed them in the most recent commit. This is ready for review.

inducer commented on examples/sod-mpi.py:
```diff
@@ -51,17 +52,36 @@
+def main(ctx_factory=cl.create_some_context, use_profiling=False, use_logmgr=False):
     """Drive the example."""
+    if use_logmgr:
+        logmgr = LogManager("sod1d.dat", "wu")  # , comm=...
```
Why not pass the communicator?

matthiasdiener replied:
Done

inducer commented on examples/sod-mpi.py:
```diff
@@ -100,11 +120,22 @@ def main(ctx_factory=cl.create_some_context):
+    if use_logmgr:
+        logmgr.add_quantity(PhysicalQuantity(discr, eos, "pressure", "P", "min"))
+        logmgr.add_quantity(PhysicalQuantity(discr, eos, "temperature", "K", "min"))
+        if rank == 0:
+            logmgr.add_watches(["step", "t_step", "min_pressure", "min_temperature"])
```
`if rank == 0` should not be necessary. Watches are [only printed on the head rank](https://github.com/inducer/logpyle/blob/99c2bd2680f27e4eaa60617fcdf89af1d15163d7/logpyle/__init__.py#L1012).

matthiasdiener replied:
Done

inducer commented on examples/sod-mpi.py:
```diff
@@ -100,11 +120,22 @@ def main(ctx_factory=cl.create_some_context):
+            logmgr.add_watches(["step", "t_step", "min_pressure", "min_temperature"])
+        if use_profiling:
+            logmgr.add_quantity(KernelProfile(actx, "diff", "flops"))
+            if rank == 0:
```
Same here.

matthiasdiener replied:
Done

inducer commented on examples/sod-mpi.py:
```diff
@@ -100,11 +120,22 @@ def main(ctx_factory=cl.create_some_context):
+            if rank == 0:
+                logmgr.add_watches(["diff_flops"])
+
+    visualizer = make_visualizer(discr, order + 3
+                                 if discr.dim == 2 else order)
```
Why the line break?

matthiasdiener replied:
Fixed.

inducer commented on examples/sod-mpi.py:
```diff
@@ -51,17 +52,36 @@
+        add_run_info(logmgr)
+        add_general_quantities(logmgr)
+        add_simulation_quantities(logmgr)
+    else:
+        logmgr = None
```
Why?

matthiasdiener replied:
logmgr is need below, as an argument to advance_state.

inducer commented on examples/vortex-mpi.py:
```diff
@@ -105,8 +127,34 @@ def main(ctx_factory=cl.create_some_context):
+            if rank == 0:
+                logmgr.add_watches(["diff_flops"])
+
+    visualizer = make_visualizer(discr, order + 3
+                                 if discr.dim == 2 else order)
```
Line break?

matthiasdiener replied:
Fixed.

inducer commented on examples/vortex-mpi.py:
```diff
@@ -52,20 +53,41 @@
 
 @mpi_entry_point
-def main(ctx_factory=cl.create_some_context):
+def main(ctx_factory=cl.create_some_context, use_profiling=False, use_logmgr=False):
```
Same suggested changes as in `sod-mpi.py`.

We can't keep repeating this stuff. Could you see about factoring out more parts of this into common functions?

matthiasdiener replied:
I have adapted the changes from sod-mpi to here, and restored the master version of sod-mpi for now until we have the API nailed down. 

- removed argument passing, replaced it with a bool config value in the driver
- passed the communicator

inducer commented on examples/sod-mpi.py:
```diff
@@ -152,9 +183,22 @@ def my_checkpoint(step, t, dt, state):
 if __name__ == "__main__":
     logging.basicConfig(format="%(message)s", level=logging.INFO)
-    main()
+    import argparse
+    parser = argparse.ArgumentParser(description="Sod (MPI version)")
+    parser.add_argument("--profile", action="store_true",
+        help="enable kernel profiling")
+    parser.add_argument("--logging", action="store_true",
+        help="enable time series logging")
+    args = parser.parse_args()
+
+    main(use_profiling=args.profile, use_logmgr=args.logging)
```
Not a fan. The driver should act as the "configuration file". Not sure I like having more "command-line configurability".

matthiasdiener replied:
Replaced with two config bools instead of argument parsing.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        p = pathlib.Path(".").resolve()
+        for d in p.parents:
+            candidate = pathlib.Path(d).joinpath("version.sh")
+            if candidate.is_file() and "emirge" in candidate.open().read():
```
Possibly expensive.

inducer replied:
Leaks a file descriptor, leading to a warning.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        mgr.set_constant("emirge_package_versions",
+            "Could not find emirge's version.sh. No package versions recorded.")
+
+    output = os.popen(path_to_version_sh).read()
```
Use `subprocess` (or add comment why it's not being used).

matthiasdiener replied:
Done.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        mgr.set_constant("emirge_package_versions",
+            "Could not find emirge's version.sh. No package versions recorded.")
+
+    output = os.popen(path_to_version_sh).read()
```
Make sure to fail if the command fails, given that you're convinced that you're talking to the right script.

matthiasdiener replied:
Done

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+
+    output = os.popen(path_to_version_sh).read()
+    if output == "":
+        output = "Could not record emirge's package versions."
+
+    mgr.set_constant("emirge_package_versions", output)
```
What data type is being recorded here? Are we separating conda package versions and pip package versions?

matthiasdiener replied:
Just plain strings for now. No separation is done currently.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+    mgr.set_constant("emirge_package_versions", output)
+
+
+def set_state(mgr: LogManager, state: ndarray) -> None:
```
"State" is not a good name. "Discretization field vector"? (here, in `StateConsumer`, etc.)

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        self.state = None
+
+    def set_state(self, state: ndarray) -> None:
+        """Set the state of the object."""
```
```suggestion
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+
+        self.quantity = quantity
+
+        if rank_aggr is None:
+            self.rank_aggr = "norm_2"
+        else:
+            self.rank_aggr = rank_aggr
+
+        from functools import partial
+
+        if op == "min":
+            self._myop = partial(self.discr.nodal_min, "vol")
+        elif op == "max":
+            self._myop = partial(self.discr.nodal_max, "vol")
+        elif op == "sum":
+            self._myop = partial(self.discr.nodal_sum, "vol")
+        else:
+            raise RuntimeError(f"unknown operation {op}")
```
Observe that `op` already entails an appropriate rank aggregator. Why use `norm2` by default when a better default is obvious?

matthiasdiener replied:
I changed it to what you suggested, I think.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        from functools import partial
+
+        if op == "min":
+            self._myop = partial(self.discr.nodal_min, "vol")
```
```suggestion
            self._discr_reduction = partial(self.discr.nodal_min, "vol")
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "mass":
+                unit = "kg"
+            elif quantity == "energy":
+                unit = "J"
+            elif quantity == "momentum":
+                if dim is None:
+                    raise RuntimeError("Missing 'dim' parameter for dimensional "
+                                       f"ConservedQuantity '{quantity}'.")
+                unit = "kg*m/s"
```
I don't think these are right.

- First, they're densities. (i.e. `/m^3`)
- Second, I'm not sure that "kg" is our fundamental mass unit. @MTCam?
- Third, it's really not the role of this code to have in-depth knowledge about physics. We should have a `FluidDynamics` object that knows this, the EOS, the dimensionality and such.

MTCam replied:
Units used in the simulation are set by the user.  They are not inherent to the code, but the user's input data.   This is all done when the user sets the EOS, and the corresponding initial state.  For example, the user may:

```
eos = IdealSingleGas(gas_constant=8.314xxxxx)
```
This setting (which is the default) is consistent with kg (mass), Pascals (pressure), Kelvin (temp), etc.   Or the user can set:
```
eos = IdealSingleGas(gas_constant=.73xxxx)
```
Such a setting would be for users who provide initial state with lb, atm, Rankine.   

The user needs to make sure they specify the gas constant R, and initial state with consistent units.  CFD users are well-accustomed to this.  The built-in solution initializers and EOS are consistent with each other so that initial states (using built-in initializers) have the proper values consistent with the units.

If we have units specified anywhere in output data, they need to be settable by the user.  One way to make it automatically consistent is to just label it with the physical quantity (i.e. mass/volume, energy/volume, momentum/volume) for now.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        if self.state is None:
+            return None
+
+        from mirgecom.euler import split_conserved
+
+        cv = split_conserved(self.discr.dim, self.state)
+
+        self.state = None
+
+        cq = getattr(cv, self.quantity)
+
+        if not isinstance(cq, DOFArray):
+            return self._myop(cq[self.dim])
+        else:
+            return self._myop(cq)
```
Why so many blank lines?

matthiasdiener replied:
Fixed.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, dim: int = None, name: str = None,
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "mass":
+                unit = "kg"
+            elif quantity == "energy":
+                unit = "J"
+            elif quantity == "momentum":
+                if dim is None:
+                    raise RuntimeError("Missing 'dim' parameter for dimensional "
+                                       f"ConservedQuantity '{quantity}'.")
+                unit = "kg*m/s"
+            else:
+                unit = ""
```
I feel that we could have a class per quantity, which would get rid of this pile of `if`s.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+
+
+class ConservedQuantity(PhysicalQuantity):
+    """Logging support for conserved quantities (mass, energy, momentum)."""
```
This code is not in charge of knowing what a conserved quantity is. Refer to the docs for `split_conserved`.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        if name is None:
+            name = f"{op}_{quantity}{dim}"
+
+        PhysicalQuantity.__init__(self, discr, quantity, unit, op, name, rank_aggr)
```
Use `super()`.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        self.state = state
+
+
+class PhysicalQuantity(LogQuantity, StateConsumer):
```
```suggestion
class DiscretizationBasedQuantity(LogQuantity, StateConsumer):
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+        raise NotImplementedError
+
+
+class ConservedQuantity(PhysicalQuantity):
```
```suggestion
class ConservedDiscretizationBasedQuantity(DiscretizationBasedQuantity):
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+
+        return self._myop(getattr(dv, self.quantity))
+
+
```
Add sectioning comments `# {{{ discretization-based quantities`, `# }}}`, etc.

inducer commented on mirgecom/profiling.py:
```diff
@@ -36,6 +36,24 @@
+nonloopy_profile_results = {}
+
+
+@dataclass(eq=True, frozen=True)
+class NonLoopyProfilekernel:
+    """Class to hold the name for a non-loopy profile result."""
+
+    name: str
```
? Explain.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+            unit = ""
+
+        if name:
+            LogQuantity.__init__(self, name,
+                                 unit, f"{stat} of '{kernel_name}'")
+        else:
+            LogQuantity.__init__(self, f"{kernel_name}_{stat}",
+                                 unit, f"{stat} of '{kernel_name}'")
```
Use `super()`.

matthiasdiener replied:
Done.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+            unit = ""
+
+        if name:
+            LogQuantity.__init__(self, name,
+                                 unit, f"{stat} of '{kernel_name}'")
+        else:
+            LogQuantity.__init__(self, f"{kernel_name}_{stat}",
+                                 unit, f"{stat} of '{kernel_name}'")
```
Avoid conditionals, especially those with duplicated code.

matthiasdiener replied:
Done.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+    num_calls, flops, bytes_accessed)."""
+
+    def __init__(self, actx: PyOpenCLProfilingArrayContext,
+                 kernel_name: str, stat: str, name: str = None):
```
Document what `stat` is.

inducer commented on examples/sod-mpi.py:
```diff
@@ -100,11 +120,22 @@ def main(ctx_factory=cl.create_some_context):
+        if rank == 0:
+            logmgr.add_watches(["step", "t_step", "min_pressure", "min_temperature"])
+        if use_profiling:
+            logmgr.add_quantity(KernelProfile(actx, "diff", "flops"))
```
I think this is impossible to use. How is a user going to know what all the relevant kernel names are? These should be added automatically.

matthiasdiener replied:
Kernels are now added automatically.

inducer commented on requirements.txt:
```diff
@@ -3,7 +3,9 @@ mpi4py
 numpy
 pytest
 pyvisfile
+pyinstrument
```
```suggestion
```

inducer commented on mirgecom/simutil.py:
```diff
@@ -131,10 +131,18 @@ def sim_checkpoint(discr, visualizer, eos, q, vizname, exact_soln=None,
         from mirgecom.io import make_rank_fname, make_par_fname
         rank_fn = make_rank_fname(basename=vizname, rank=rank, step=step, t=t)
+        if vis_timer:
+            import time
+            start = time.time()
+
         visualizer.write_parallel_vtk_file(
             comm, rank_fn, io_fields, overwrite=overwrite,
             par_manifest_filename=make_par_fname(basename=vizname, step=step, t=t))
 
+        if vis_timer:
+            stop = time.time()
+            vis_timer.add_time(stop-start)
+
```
Use the subtimer context manager for this.

matthiasdiener replied:
Done.

inducer commented on mirgecom/profiling.py:
```diff
@@ -135,7 +208,8 @@ def tabulate_profiling_data(self) -> pytools.Table:
                 fprint_max = "--"
 
-            bytes_per_flop = [f / b for f, b in zip(flops, bytes_accessed)]
+            bytes_per_flop = [f / b if b > 0 else 0
```
```suggestion
            bytes_per_flop = [f / b if b > 0 else None
```
?

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+    """Logging support for physical quantities."""
+
+    def __init__(self, discr: Discretization, quantity: str, unit: str, op: str,
+                 name: str, rank_aggr: str = None):
```
Why is `rank_aggr` a constructor parameter? (When it only sets the default, which can be overridden by the user when querying?)

matthiasdiener replied:
Removed.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,247 @@
+                 quantity: str, op: str, unit: str = None, name: str = None,
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "temperature":
+                unit = "K"
+            elif quantity == "pressure":
+                unit = "P"
+            else:
+                unit = ""
```
Same as above. This code shouldn't need to know about units. It should ask someone else.

matthiasdiener replied:
I don't disagree, but I don't think the functionality is there yet.

matthiasdiener commented on examples/vortex-mpi.py:
```diff
@@ -52,20 +53,45 @@
+        add_general_quantities(logmgr)
+        add_simulation_quantities(logmgr)
+    else:
+        logmgr = None
```
`logmgr` is used below, so have to assign a value here.

matthiasdiener commented on examples/vortex-mpi.py:
```diff
@@ -52,20 +53,45 @@
+
+    if use_logmgr:
+        logmgr = LogManager("vortex.dat", "wu", mpi_comm=comm)
+        add_run_info(logmgr)
+        add_package_versions(logmgr)
+        add_general_quantities(logmgr)
+        add_simulation_quantities(logmgr)
```
I was thinking of adding this into a new class `MirgecomLogManager` or so. What do you think?

Also, why are some functions methods of the class, but some are global functions, e.g. 

`add_run_info(logmgr)`, but `logmgr.add_quantity(...)`

matthiasdiener reviewed (commented):
Thank you for the detailed review, @inducer ! I think I addressed most of your comments and think this is ready for another review.

inducer commented:
Let me know when the units warning and the monkeypatch thing are done. I'll take another look then.

matthiasdiener commented:
> Let me know when the units warning and the monkeypatch thing are done. I'll take another look then.

Done - all monkeys patched ;-)

matthiasdiener commented:
Added logging of the device name. This is ready for review.

matthiasdiener commented:
Added memory profiling. Ready for review @inducer .

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,230 @@
+
+    if os.path.isfile("../version.sh"):
+        stream = os.popen("../version.sh")
+        mgr.set_constant("package_versions", stream.read())
```
Done

matthiasdiener commented on mirgecom/profiling.py:
```diff
@@ -31,11 +31,59 @@
+
+nonloopy_profile_results = {}
+
+
```
Also monkey-patch add_events

matthiasdiener replied:
Done, ready for review @inducer .

inducer commented on mirgecom/profiling.py:
```diff
@@ -31,11 +31,96 @@
+nonloopy_profile_results = {}
+
+
+def add_nonloopy_pyopencl_results(queue, events: list) -> None:
+    """Gather and store pyopencl.array results."""
+    if queue and \
+      queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+
+        times = []
+        for evt in events:
+            time = evt.profile.end - evt.profile.start
+            times.append(time)
+
+        add_nonloopy_profiling_result("pyopencl_array", mean(times))
+
+
+def pyopencl_monkey_del(self):
+    """Monkey patches ``pyopencl.array.Array.__del__``  to grab profile data."""
+    if self.events:
+        cl.wait_for_events(self.events)
+
+        add_nonloopy_pyopencl_results(self.queue, self.events)
+
+        del self.events[:]
```
This should preserve and call the old `__del__` instead of reimplementing it. Same for the other methods.

inducer replied:
I think we'll actually get better/higher-fidelity information if we monkey-patch `pyopencl.array.elwise_kernel_runner`. At least we'll be able to get bandwidth info by assuming that each array involved gets read/written once.

inducer commented on mirgecom/profiling.py:
```diff
@@ -216,6 +371,16 @@ def _get_kernel_stats(self, program, kwargs: dict) -> ProfileResult:
 
             self.kernel_stats.setdefault(program, {})[args_tuple] = res
+
+            init_monkey_patch()
```
The monkeypatch should be reversed upon destruction of the array context.

inducer commented on mirgecom/profiling.py:
```diff
@@ -31,11 +31,96 @@
+
+@dataclass(eq=True, frozen=True)
+class NonLoopyProfilekernel:
+    """Class to hold the name for a non-loopy profile result.
+
+    This is necessary so that :meth:`tabulate_profiling_data` etc. can
+    access the 'name' field of the non-Loopy kernel.
```
Doc: where and by who is this used?

inducer commented on mirgecom/profiling.py:
```diff
@@ -94,9 +181,54 @@ def _finish_profile_events(self) -> None:
+
+                bytes_accessed += [v.bytes_accessed / 1e9
+                              if v.bytes_accessed is not None else 0 for v in value]
+                # fprint_bytes += [v.footprint_bytes if v.footprint_bytes
+                #                  is not None else 0 for v in value]
```
?

inducer commented on mirgecom/profiling.py:
```diff
@@ -94,9 +181,54 @@ def _finish_profile_events(self) -> None:
+        num_calls, times, flops, bytes_accessed = \
+            _gather_data(self.profile_results)
+
+        if num_calls == 0:
```
This has a high risk for being incorrect. What if an elwise kernel and a loopy kernel happen to be called the same thing?

inducer commented on mirgecom/profiling.py:
```diff
@@ -94,9 +181,54 @@ def _finish_profile_events(self) -> None:
+
+        if num_calls == 0:
+            num_calls, times, flops, bytes_accessed = \
+                _gather_data(nonloopy_profile_results)
```
I'm not sure I care for this design much: It requires multiple traversals of the `*profile_results` lists, one per kernel.

inducer reviewed (commented):
Thanks for working on this, and sorry about the delay in getting this looked at. Some comments below.

inducer commented on doc/running/profiling.rst:
```diff
@@ -12,3 +16,15 @@ In addition to using this array context, you also need to enable profiling in th
+
+.. automodule:: mirgecom.logging_quantities
+
+An overview of how to use logpyle is given `here <https://logpyle.readthedocs.io/en/latest/log.html#how-do-i-use-this>`_.
```
FYI: You can link to sections in other docs using Intersphinx, too. https://docs.readthedocs.io/en/stable/guides/intersphinx.html

matthiasdiener replied:
Done

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,350 @@
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "sum":
+            self._discr_reduction = partial(self.discr.nodal_sum, "vol")
```
Added norm(2). Integrals might need an addition in grudge/eager ?

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,350 @@
+        self.state = None
+        cq = getattr(cv, self.quantity)
+
+        if not isinstance(cq, DOFArray):
```
Clarified this a bit.

matthiasdiener reviewed (commented):
Ready for another review @inducer . I think I addressed most of your comments, except:

- elwise_kernel_runner
- integral op

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,350 @@
+    def add_device_name(self, queue: cl.CommandQueue):
+        """Add the device name to the log."""
+        self.set_constant("device_name",
+                     str(queue.get_info(cl.command_queue_info.DEVICE)))
```
Done

MTCam commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,350 @@
+                if dim is None:
+                    raise RuntimeError("Missing 'dim' parameter for dimensional "
+                                       f"ConservedQuantity '{quantity}'.")
+                unit = "kg*m/s/m^3"
```
Sorry, Matthias has just directed me to this comment.   These are ok for defaults, but the user will need a way tot set them.

matthiasdiener commented:
> Ready for another review @inducer . I think I addressed most of your comments, except:
> 
> * elwise_kernel_runner
> * integral op

Also, can we do these in a separate PR? This PR has already grown to an unmanageable size, and GitHub doesn't even render the page correctly anymore.

matthiasdiener commented:
> > Ready for another review @inducer . I think I addressed most of your comments, except:
> > 
> > * elwise_kernel_runner
> > * integral op
> 
> Also, can we do these in a separate PR? This PR has already grown to an unmanageable size, and GitHub doesn't even render the page correctly anymore.

Added a new issue to track these two (#186). Removed the `sum` op from CV. Ready for another review @inducer.

matthiasdiener commented:
Switched to https://github.com/inducer/pyopencl/pull/415 - this fixed the 'phantom' extra calls to the elwise kernels. Kernel counts now match 100% with the nvprof counts for the cases I've tried.

matthiasdiener commented:
Most of the profiling-related changes are now in #192 .

matthiasdiener commented:
Cleaned up the latest commits. Ready for review @inducer .

matthiasdiener commented:
Set it to log unitless by default, as per our discussion on 02/05. Ready for review @inducer.

MTCam commented on examples/vortex-mpi.py:
```diff
@@ -108,6 +127,26 @@ def main(ctx_factory=cl.create_some_context):
+        logmgr_add_device_name(logmgr, queue)
+        logmgr_add_default_discretization_quantities(logmgr, discr, dim)
+
+        logmgr.add_watches(["step.max", "t_step.max", "t_log.max",
+                            "min_temperature", "norm_momentum1"])
```
Can these be changed easily to temperature, pressure (min, max) for both?

matthiasdiener replied:
Yes, you would need to call `logmgr.add_watches(["min_temperature", "max_temperature", "min_pressure", "max_pressure"])`

MTCam replied:
Doing that would provide a more relevant example, but I realize that it would then not be demonstrative of the loggers facility for handling vector quantities.

MTCam commented on mirgecom/euler.py:
```diff
@@ -352,6 +352,32 @@ def get_inviscid_cfl(discr, eos, dt, q):
     return dt / wanted_dt
 
 
+# By default, run unitless
+NAME_TO_UNITS = {
+    "mass": "",
+    "energy": "",
+    "momentum": "",
+    "temperature": "",
+    "pressure": ""
+}
+
+
+def units_for_logging(quantity: str) -> str:
+    """Return unit for quantity."""
+    return NAME_TO_UNITS[quantity]
```
I see why we're doing this for our project.  I question whether the logger utility really needs this interface, however.  Can't the user define the units when they identify watched quantities?  

This could be misguided... but IMO, it is outside the scope of the logger to have knowledge of the physical content of the quantities it is watching.

matthiasdiener replied:
Sure, a user would just need to define their own `units_for_logging` function and pass that to the initialization. See e.g. https://github.com/illinois-ceesd/mirgecom/blob/ff9e2ea031794940fbf1352273489eebadc39ca0/examples/vortex-mpi.py#L72-L74

MTCam commented on mirgecom/euler.py:
```diff
@@ -352,6 +352,32 @@ def get_inviscid_cfl(discr, eos, dt, q):
+def extract_vars_for_logging(dim: int, state: np.ndarray, eos) -> dict:
+    """Extract state vars."""
+    cv = split_conserved(dim, state)
+    dv = eos.dependent_vars(cv)
+
```
Depending on which EOS - this call can be quite expensive.

MTCam commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+    def __init__(self):
+        self.state_vars = None
+
+    def set_sim_state(self, dim, state, eos) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = extract_state_vars(dim, state, eos)
```
Related to previous comment about `eos.dependent_vars` - we just need to be careful because this can get expensive.

MTCam commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.dim is not None:  # momentum
+            return self._discr_reduction(quantity[self.dim])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
```
Will this break when we have other non-scalar quantities (e.g. `scalar_mass` for mixtures) who's dimension is not spatially-dependent?

matthiasdiener replied:
I'm not sure I understood correctly, but I think this will not break. In this context `self.dim` isn't necessarily related or restricted to the number of dimensions of the input problem. It might be better to rename that variable to something like `q_dim` or so.

MTCam replied:
Yup, perfect!  Thanks!

MTCam reviewed (commented):
Thanks!  Just a few thoughts...

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+    def set_sim_state(self, dim, state, eos) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = extract_state_vars(dim, state, eos)
```
Additionally, this call occurs once per quantity, not once for all of them. (which should be easy to fix)

matthiasdiener replied:
Implemented a (hacky) fix.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+    def __call__(self) -> float:
+        """Return the memory usage."""
+        from memory_profiler import memory_usage  # pylint: disable=import-error
+        return memory_usage(-1, 0)[0]
```
Look at [what this does under the hood](https://github.com/pythonprofilers/memory_profiler/blob/685b5f8db2a838cd324e0d686cdf7f5290281706/memory_profiler.py#L232-L437). Do you really want all that to happen on every time step? Note the timing while loops. We can probably just use `psutil` directly: https://github.com/pythonprofilers/memory_profiler/blob/685b5f8db2a838cd324e0d686cdf7f5290281706/memory_profiler.py#L126-L143

matthiasdiener replied:
good point, changed to psutil.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+        names = [f"{kernel_name}_time", f"{kernel_name}_flops",
+                 f"{kernel_name}_num_calls", f"{kernel_name}_bytes_accessed",
+                 f"{kernel_name}_footprint"]
```
Disagree with hardcoding this. Make the output of `get_profiling_data_for_kernel` a `dataclass`, and walk its fields.

matthiasdiener replied:
Done.

inducer commented on mirgecom/profiling.py:
```diff
@@ -155,9 +160,51 @@ def _finish_profile_events(self) -> None:
 
-    def tabulate_profiling_data(self) -> pytools.Table:
+    def get_profiling_data_for_kernel(self, kernel_name: str,
+                                      wait_for_events=True) -> float:
```
That return type is wrong AFAICT.

matthiasdiener replied:
Fixed.

inducer commented on mirgecom/profiling.py:
```diff
@@ -315,6 +362,14 @@ def _cache_kernel_stats(self, program: lp.kernel.LoopKernel, kwargs: dict) \
+
+            if self.logmgr:
+                if "pyopencl_array_time" not in self.logmgr.quantity_data:
+                    self.logmgr.add_quantity(KernelProfile(self, "pyopencl_array"))
```
Missed it earlier, I'll nag about it here: These should be identified by their names.

matthiasdiener replied:
This will be fixed automatically with #193.

matthiasdiener replied:
Should be fixed now.

inducer reviewed (commented):
Sorry about the long delay. Finally had some time to take a look.

matthiasdiener commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ numpy
 pytest
 pyvisfile
 pymetis
+logpyle
```
No need imho, but we can change it if you prefer.

inducer replied:
Fair. Let's leave it.

MTCam commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim):
+    """Add default discretization quantities to the logmgr."""
+    for quantity in ["pressure", "temperature"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+    for quantity in ["mass", "energy"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, dim=dim))
```
Usually, people want to watch pressure, temperature, the velocities (each component), mach number, speed of sound, CFL, infrequently the mass. Reaction rates and mass fractions are frequently watched - but only certain components of it.

Pressure and temperature seem like harmless defaults if there need to be defaults. Pressure only actually would be fine for a lot of runs.  

If these fields are easily set and controlled at the driver level, then I'm not sure I'd set more than pressure as a default and allow to be easily disabled or replaced.

MTCam commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim):
+    """Add default discretization quantities to the logmgr."""
+    for quantity in ["pressure", "temperature"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+    for quantity in ["mass", "energy"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, dim=dim))
```
I keyed on this too, but isn't the `logging_quantities.py` on the _MIRGE-Com_ side of setting up logging quantities?  If that is true, then I think this is OK from that perspective. 

If this is indeed part of the logging infrastructure itself, then I agree that the logger shouldn't be in the business of knowing much about the simulation-side content of the logged quantities. 

What might be troublesome for users is if the logged quantities aren't easily set up (i.e. which fields, and supporting derived fields) from run-to-run.  While I think it is ok if the user needs to modify a file to control their logged quantities, I'm not sure that down in _MIRGE-Com_ library is where that file should live. 

Is it possible we need a place to put utilities that are outside the _MIRGE-Com_ "library" but part of the infrastructure used to build drivers?

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+    logmgr = LogManager(filename=filename, mode=mode, mpi_comm=mpi_comm)
+
+    global extract_state_vars, units_logging
+    extract_state_vars = extract_vars_for_logging
+    units_logging = units_for_logging
```
Fixed.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+
+        from functools import partial
+
+        if op not in ["min", "max", "norm"]:
+            raise ValueError("op must be one of 'min', 'max', 'norm'.")
```
removed extra check.

matthiasdiener commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,306 @@
+    """
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, name: str = None, dim: int = None):
```
Changed to axis.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,327 @@
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
+    """
+    state_vars = None
```
Make this a dictionary: `extract_vars_func` -> extracted state.

matthiasdiener replied:
Done.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,327 @@
+        super().__init__(name, "MByte", description="Memory usage (RSS, host)")
+
+        # Make sure the psutil module is available
+        import importlib
+        found = importlib.util.find_spec("psutil")
```
Why not the simple `try: import` scheme? Add a comment

matthiasdiener replied:
Fixed.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,318 @@
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
```
```suggestion
    """Base class for quantities that require a state for logging.
    
    .. automethod:: __init__
    """
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,318 @@
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self, extract_vars_for_logging):
+        self.extract_state_vars = extract_vars_for_logging
```
Add docstring. Give context on what `extract_vars_for_logging` is supposed to do.

matthiasdiener replied:
Added.

matthiasdiener replied:
Added

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,318 @@
+        from mirgecom.profiling import PyOpenCLProfilingArrayContext
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+
```
```suggestion
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,318 @@
+
+        for f in fields(ProfileResultsForKernel):
+            names.append(f"{kernel_name}_{f.name}")
+            if f.name == "num_calls":
+                units.append("1")
+            elif f.name == "flops":
+                units.append("GFlops")
+            elif f.name == "time":
+                units.append("s")
+            elif f.name == "bytes_accessed" or f.name == "footprint_bytes":
+                units.append("GByte")
+            else:
+                raise RuntimeError(f"unknown unit for field {f.name}")
```
This wants to be a dictionary and a list comprehension that accesses it, I think.

matthiasdiener replied:
Changed.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,318 @@
+class PythonMemoryUsage(LogQuantity):
+    """Logging support for Python memory usage (RSS, host).
+
+    Uses :mod:`psutil` to track memory usage. Virtually no overhead.
```
This is missing from the docs. (Otherwise the unresolved reference to the `psutil` docs would cause the doc build to fail.)

matthiasdiener replied:
Added

inducer commented on mirgecom/profiling.py:
```diff
@@ -31,52 +31,16 @@
-
-
 @dataclass
 class ProfileResult:
```
```suggestion
class SingleCallKernelProfile:
```

inducer commented on mirgecom/profiling.py:
```diff
@@ -87,26 +51,15 @@ class ProfileResult:
-
-    return evt
+@dataclass
+class ProfileResultsForKernel:
```
```suggestion
class MultiCallKernelProfile:
```

inducer reviewed (commented):
Some more thoughts below, mostly minor.

inducer commented on mirgecom/profiling.py:
```diff
@@ -168,18 +149,57 @@ def _finish_profile_events(self) -> None:
-        nonloopy_profile_events = []
 
-    def tabulate_profiling_data(self) -> pytools.Table:
+    def get_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> ProfileResultsForKernel:
+        """Return value of profiling result for kernel `kernel_name`."""
```
I'm not sure I like that this function resets the counters. At the very least, that behavior should be documented here.

inducer replied:
I'm not sure I like that this function resets the counters. At the very least, that behavior should be documented here.
```suggestion
    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
                                 wait_for_events=True) -> ProfileResultsForKernel:
        """Return value of profiling result for kernel `kernel_name` and reset
        the internal profiling data for this kernel."""
```

inducer commented on mirgecom/profiling.py:
```diff
@@ -168,18 +149,57 @@ def _finish_profile_events(self) -> None:
+        if wait_for_events:
+            self._finish_profile_events()
+
+        def _gather_data(results_list: list):
```
That type annotation seems wrong.
```
del results_list[key]
```
wouldn't work on a list.

(And if it's not a list, it shouldn't be called "list" either.)

matthiasdiener replied:
Fixed.

matthiasdiener commented:
Thanks @inducer . I think this is ready for another review.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim,
+      extract_vars_for_logging, units_for_logging):
+    """Add default discretization quantities to the logmgr."""
+    for quantity in ["pressure", "temperature"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op, extract_vars_for_logging, units_for_logging))
+    for quantity in ["mass", "energy"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op, extract_vars_for_logging, units_for_logging))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, extract_vars_for_logging, units_for_logging,
+                axis=dim))
```
Computing all those will be quite expensive, since it's lots of reductions (the GPU kind). I'm not sure we should do that by default. Maybe allow the user to specify a list of strings for what they actually want, and use a more parsimonious default.

matthiasdiener replied:
Note that these aren't activated by default. Maybe just rename the function?

inducer replied:
Yeah, that would help.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+                discr, quantity, op, extract_vars_for_logging, units_for_logging))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
```
Those `for op` loops are pretty redundant. Flip the nesting and use only one of them?

matthiasdiener replied:
Done.

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
```
I would prefer `L2_norm` to `norm`, to clarify. (`L2` is different mathematically from `l2`, hence the capital letter, which I otherwise wouldn't use.)

matthiasdiener replied:
changed to L2_norm

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
```
```suggestion
            self._discr_reduction = partial(self.discr.norm, p=2)
```
(Otherwise, why is that `partial` there?)

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.axis is not None:  # momentum
+            return self._discr_reduction(quantity[self.axis])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
```
```suggestion
        if self.axis is not None:  # e.g. momentum
            quantity = quantity[self.axis]

        return self._discr_reduction(quantity)
```
(Less duplicated logic.)

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
```
```suggestion
        The :class:`logpyle.LogManager` whose :class:`StateConsumer` quantities will receive *state*.
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,319 @@
+        Parameters
+        ----------
+        extract_vars_for_logging
+            Returns a dict of the state vars for a particular state.
```
What are the arguments? What are the types of key and value of that dict?

matthiasdiener replied:
Added

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +93,115 @@ def __init__(self, queue, allocator=None) -> None:
+            self._finish_profile_events()
+
+        def _gather_data(results_dict: dict):
+            results = [key for key in results_dict if hasattr(key, "name")
+                       and key.name == kernel_name]
+            results.extend(
+                [key for key in results_dict if hasattr(key, "function_name")
+                 and key.function_name == kernel_name])
```
These traversals could get quite expensive. Why not insert them into the dictionary under their name in the first place if that's all we care about in the end?

matthiasdiener replied:
I've simplified the code a bit; the reason not to index them by name is that sometimes we want to filter by name+kwargs

inducer replied:
Not sure I understand. Could you give an example?

matthiasdiener replied:
In `tabulate_profiling_data()`, we separate kernels based on the actual kernel object, as well as the `args_tuple` (=`kwargs`) that the kernel was called with in order to differentiate between different kernels that might have the same name. In the logging infrastructure, this is impractical, and we only look at the kernel name.

I've simplified and improved the code for `get_and_reset_profiling_data_for_kernel` a bit more.

inducer reviewed (approved):
A few more. Good to go after these.

inducer commented on mirgecom/profiling.py:
```diff
@@ -110,6 +110,7 @@ def array_kernel_exec_hook(self, knl, queue, gs, ls, *actual_args, wait_for):
         evt = knl(queue, gs, ls, *actual_args, wait_for=wait_for)
 
         name = knl.function_name
+        knl.name = knl.function_name  # "name" is used by the LoopKernels
```
Not a fan of modifying `cl.Kernel` objects. IMO this is not an improvement.

matthiasdiener replied:
I undid this.

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+            for arg in actual_args if isinstance(arg, cl.array.Array))
+
+        try:
+            ignore = self.kernel_stats[knl][args_tuple]  # noqa
```
```suggestion
            self.kernel_stats[knl][args_tuple]
```

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+
+        return evt
 
     def _finish_profile_events(self) -> None:
```
IMO this method is a bit mis-named: It transfers data from `profile_events` to `profile_results`. Maybe a different name could be found that reflects that better?

matthiasdiener replied:
renamed to `_wait_and_transfer_profile_events`

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> MultiCallKernelProfile:
+        """Return and reset profiling data for kernel `kernel_name`."""
+        if wait_for_events:
+            self._finish_profile_events()
```
Why is `wait_for_events` a good argument name for "would you like to work with up-to-date data"?

matthiasdiener replied:
See comment about  removing this option below.

matthiasdiener replied:
removed the option.

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> MultiCallKernelProfile:
+        """Return and reset profiling data for kernel `kernel_name`."""
+        if wait_for_events:
+            self._finish_profile_events()
```
Also, document `wait_for_events`.

matthiasdiener replied:
I was thinking of removing the option completely. Is `wait_for_events=False` ever useful?

matthiasdiener replied:
removed the option.

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+            if value[0].footprint_bytes is not None:
+                fprint_bytes += sum([v.footprint_bytes for v in value])
+
+            del self.profile_results[key]
```
Is this essential? Why do we need to consume/zero out `profile_results` (instead of letting it grow)? (If we'd like that functionality, could we make it a separate method?)

matthiasdiener replied:
I think the profile data will consume a lot of memory if we don't delete the old one. We would also need a mechanism to avoid getting stale data from a previous timestep (also to avoid iterating over longer and longer lists).

inducer replied:
It is *possible* to separate the "get" from the "reset" functionality while avoiding long lists: We could simply retain the computed `MultiCallKernelProfile` in `profile_results` and incorporate it/use it as a starting point in the next round of statistics. The bigger question is: do we want that? I could go both ways here.

matthiasdiener replied:
> It is _possible_ to separate the "get" from the "reset" functionality while avoiding long lists: We could simply retain the computed `MultiCallKernelProfile` in `profile_results` and incorporate it/use it as a starting point in the next round of statistics. The bigger question is: do we want that? I could go both ways here.

Hmm, not sure I understand. In the current implementation, a MultiCallKernelProfile object holds (mostly averaged) data for multiple executions of a single kernel during a single time step. Should the MultiCallKernelProfile span more than one time step? Or in other words, how would it be incorporated into the next round of statistics?

inducer replied:
> MultiCallKernelProfile object holds (mostly averaged) data for multiple executions of a single kernel during a single time step. Should the MultiCallKernelProfile span more than one time step

First off, this array context has no idea what a time step is, so the period over which statistics are kept is ultimately "from one reset to the next". I'm pushing for separating "reset" from "get statistics".

> Or in other words, how would it be incorporated into the next round of statistics?

If you have an average representing 10 data points, and you add one data point, you compute a 10:1 weighted average to compute the new average over 11 data points. In that sense, a `MultiCallKernelProfile` could simply represent multiple calls and could be incorporated into a new average by weighting.

matthiasdiener replied:
> First off, this array context has no idea what a time step is, so the period over which statistics are kept is ultimately "from one reset to the next". I'm pushing for separating "reset" from "get statistics".

Ok, fair. How would the "reset" get called from the logging infrastructure?

> > Or in other words, how would it be incorporated into the next round of statistics?
> 
> If you have an average representing 10 data points, and you add one data point, you compute a 10:1 weighted average to compute the new average over 11 data points. In that sense, a `MultiCallKernelProfile` could simply represent multiple calls and could be incorporated into a new average by weighting.

Ok.

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+
+        keys = [k for k in self.profile_results.keys() if k.name == kernel_name]
+
+        for key in keys:
+            value = self.profile_results[key]
```
`key` and `value` are not useful variable names IMO. They contain no information as to what object type you're dealing with.

matthiasdiener replied:
Good point, fixed here and below.

inducer commented on mirgecom/profiling.py:
```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+        bytes_accessed = 0
+        fprint_bytes = 0
+
+        keys = [k for k in self.profile_results.keys() if k.name == kernel_name]
```
Continued from https://github.com/illinois-ceesd/mirgecom/pull/96#discussion_r578866021:

This isn't better IMO; if anything, it's slower because the list of kernel-things it traverses is longer. Lookup by name should be O(1), otherwise you're using the wrong data structure.

I looked, I wasn't able to find any code that uses anything other than the kernel name. Could you point me to it?

matthiasdiener replied:
> Continued from [#96 (comment)](https://github.com/illinois-ceesd/mirgecom/pull/96#discussion_r578866021):
> 
> This isn't better IMO; if anything, it's slower because the list of kernel-things it traverses is longer. Lookup by name should be O(1), otherwise you're using the wrong data structure.

`len(self.profile_results.keys())` should be pretty short, say O(10). `len(keys)` should be in the worst case O(3). The current implementation is about 20% faster of what I had before.

> I looked, I wasn't able to find any code that uses anything other than the kernel name. Could you point me to it?

It's used in `_get_profiling_stats()` directly:

https://github.com/illinois-ceesd/mirgecom/blob/ce51d324754b94e26b0783dd635cd97d99bdcc69/mirgecom/profiling.py#L267-L269

and then the result of that is used indirectly in `tabulate_profiling_data()`:

https://github.com/illinois-ceesd/mirgecom/blob/ce51d324754b94e26b0783dd635cd97d99bdcc69/mirgecom/profiling.py#L201

 `tabulate_profiling_data()`, in contrast to the logging of kernel profiling data, *does* differentiate between different kernels and kernel arguments when presenting results. So we could invert the way results are stored and index them by kernel name, speeding up the logging while slowing down `tabulate_profiling_data()` (or just not differentiating between kernels of the same name). Imho, that would be a fair tradeoff, but changing this might be a big enough change for a new PR.

inducer replied:
Can you explain the benefit of differentiating beyond the name? (/provide an example of that differentiation) I still don't really understand what you mean here.

matthiasdiener replied:
> Can you explain the benefit of differentiating beyond the name? (/provide an example of that differentiation) I still don't really understand what you mean here.

I think there are two main reasons:

1. In practical terms, it means that kernels could be analyzed separately based on their input parameters, e.g. in the following profiling output for the `grudge_assign_0` kernel: 

Function            | Calls | Time_sum [s] | Time_min [s] | Time_avg [s] | Time_max [s] | GFlops/s_min | GFlops/s_avg | GFlops/s_max | BWAcc_min [GByte/s] | BWAcc_mean [GByte/s] | BWAcc_max [GByte/s] | BWFoot_min [GByte/s] | BWFoot_mean [GByte/s] | BWFoot_max [GByte/s] | Intensity (flops/byte)
:-------------------|:------|:-------------|:-------------|:-------------|:-------------|:-------------|:-------------|:-------------|:--------------------|:---------------------|:--------------------|:---------------------|:----------------------|:---------------------|:-----------------------
grudge_assign_0     | 2     | 0.0001233    | 4.629e-05    | 6.163e-05    | 7.696e-05    | 1.029        | 1.37         | 1.711        | 6.081               | 8.096                | 10.11               | 0.00018              | 0.00018               | 0.00018              | 0.1692
grudge_assign_0     | 8     | 0.0004143    | 3.341e-05    | 5.179e-05    | 6.104e-05    | 0.118        | 0.1444       | 0.2155       | 1.769               | 2.166                | 3.233               | 0.000108             | 0.000108              | 0.000108             | 0.06667
grudge_assign_0     | 4     | 0.000175     | 3.676e-05    | 4.376e-05    | 5.955e-05    | 0.1209       | 0.1707       | 0.1958       | 1.814               | 2.56                 | 2.938               | 0.000108             | 0.000108

If I remember correctly, this was a feature you had requested.

2. Differentiating by name could lead to confusion in case two different kernels happen to have the same name (like `diff` in grudge some days ago).

Imho, neither of these two reasons is compelling enough to keep the current system in place. Apart from the performance, there is also a slight annoyance of the output not matching the e.g. nvprof output in terms of the number of different kernels.

inducer replied:
> analyzed separately based on their input parameters

I'm still not sure I understand. How are these differentiated? By the shape of the arguments? By the pointers? By the actual values in the array?

matthiasdiener replied:
> > analyzed separately based on their input parameters
> 
> I'm still not sure I understand. How are these differentiated? By the shape of the arguments? By the pointers? By the actual values in the array?

Input parameters are differentiated by the shape of array args and values of scalar args. See e.g.

https://github.com/illinois-ceesd/mirgecom/blob/785efc06c40da0a1ba9b708248dc3f9211ca7be8/mirgecom/profiling.py#L292-L298

This is then used in the result collection:

https://github.com/illinois-ceesd/mirgecom/blob/785efc06c40da0a1ba9b708248dc3f9211ca7be8/mirgecom/profiling.py#L143-L151

inducer replied:
I see. Thanks! I guess I vaguely remember pushing for this, but on second thought I'm no longer fond of this. If there are important distinctions between kernels, then we can probably just encode them in the name. Seeing the "same" kernel listed five times (but not really knowing which is which) isn't all that helpful after all. How about name-only after all?

matthiasdiener replied:
Changed to indexing by name only. Did you see my other question regarding the deletion @inducer?

> How would the "reset" get called from the logging infrastructure?

inducer replied:
> How would the "reset" get called from the logging infrastructure?

Likely the `LogQuantity` would reset the stats right after reading them. So pretty similar to before, just not forcibly connected.

matthiasdiener replied:
Ok, I have implemented that.

matthiasdiener commented:
I fixed most of the comments that you had @inducer. For two things I will need some more guidance:
- the delete issue
- the kernel results traversal

inducer commented:
> For two things I will need some more guidance:

Commented inline.

inducer commented:
To cut down on noise in my email, I'm unsubscribing from this PR. When it next needs my attention, please @-mention me or hit the "request review" button. Otherwise, I may not see your messages in a timely manner.

matthiasdiener commented:
I think I addressed all your comments @inducer. Ready for another review.

inducer commented on mirgecom/profiling.py:
```diff
@@ -123,66 +76,138 @@ class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+
+        time = sum([v.time for v in knl_results]) / num_calls / 1e9
+
+        if knl_results[0].flops is not None:
```
```suggestion
        # otherwise, the key would not exist in self.profile_results
        assert knl_results

        if knl_results[0].flops is not None:
```

matthiasdiener replied:
I don't understand this suggestion @inducer. Isn't the `assert` here superfluous with the `if kernel_name not in self.profile_results:` in line 170? If not, why not put the assert before the determination of `num_calls`?

inducer replied:
I added it because there was an access to `knl_results[0]` which isn't generically safe. (It's a pattern that my brain just picks out, I guess.) I suggested adding this so that next time I spot it, I don't have to spend time convincing myself over again that it's safe. I'm all in favor of moving it to a more sensible spot.

matthiasdiener replied:
This change isn't necessary anymore since the code uses StatisticsAccumulator now.

inducer commented on mirgecom/profiling.py:
```diff
@@ -253,7 +278,7 @@ def tabulate_profiling_data(self) -> pytools.Table:
                 bandwidth_access_mean = "--"
                 bandwidth_access_max = "--"
```
This code could probably just use `get_profiling_data_for_kernel` now, to avoid code duplication.

matthiasdiener replied:
I'm not sure this would be too useful. `get_profiling_data_for_kernel` only returns averages, while `tabulate_profiling_data` also calculates min/max as well as derived statistics such as bandwidth etc. Imho we wouldn't gain much by using `get_profiling_data_for_kernel` here.

inducer replied:
I mean, that's kind of a strange inconsistency, too. Why can one gather max/min, the other not?

matthiasdiener replied:
We can modify `get_profiling_data_for_kernel` to also log min/max, if you prefer.

inducer replied:
How about returning something like this:
```python
class StatisticsAccumulator:
    def __init__(self):
        self.num_values = 0
        self._sum = 0
        self._min = 0
        self._max = 0

    def add_value(self, v):
        if v is None:
             return
        self.num_values += 1
        self._sum += v
        # ...

    def mean(self):
        if self.num_values == 0:
            return None

        return self.sum / self.num_values
# ...
```

matthiasdiener replied:
Just as a side note, I think my original idea for just doing averages here was to reduce the sheer number of statistics and the associated clutter (+overhead). Just adding min/max would triple the amount of kernel profiling data, which is already by far the majority of data stored for each time step. Also, it would create nice-looking entries like `multiply_time_min.max`.

inducer replied:
Sure! I'm not saying we have to create `LogQuantity`s for all of them. This accumulator thing is just so that we can unify the summation logic.

matthiasdiener replied:
Added the solution you outlined. This works a bit differently for the derived quantities

inducer commented on mirgecom/profiling.py:
```diff
@@ -123,66 +76,138 @@ class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+
+    def reset_profiling_data_for_kernel(self, kernel_name: str) -> None:
+        """Reset profiling data for kernel `kernel_name`."""
+        if kernel_name in self.profile_results:
+            del self.profile_results[kernel_name]
```
```suggestion
        self.profile_result.pop(kernel_name, None)
```

inducer commented on mirgecom/logging_quantities.py:
```diff
@@ -0,0 +1,321 @@
+    logmgr.set_constant("cl_device_name", str(queue.device))
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim,
```
```suggestion
def logmgr_add_all_discretization_quantities(logmgr: LogManager, discr, dim,
```

matthiasdiener replied:
The quantities added here aren't necessarily all discretization related quantities, so renaming this is maybe confusing.

inducer replied:
Agreed. `add_many`?

matthiasdiener replied:
Could be, or `add_basic`?

inducer replied:
Don't like `add_basic` so much.

matthiasdiener replied:
renamed to `add_many`.

inducer reviewed (commented):
Alright, getting close. These three, plus the doc CI failure, and then this is good to go IMO.

matthiasdiener commented:
Ready for another review @inducer.

inducer commented on mirgecom/profiling.py:
```diff
@@ -51,15 +52,68 @@ class SingleCallKernelProfile:
+        import sys
+        self.num_values = 0
+        self._sum = 0
+        self._min = sys.maxsize
+        self._max = 0
```
I think it's better to start with `None`, to avoid assuming bounds.

matthiasdiener replied:
Done.

inducer replied:
Doesn't look done. Forgot to push?

matthiasdiener replied:
Yep, sorry. Should be pushed now.

inducer commented on mirgecom/profiling.py:
```diff
@@ -51,15 +52,68 @@ class SingleCallKernelProfile:
+    """Class that provides statistical functions for profile results.
+
+    Values are converted to "Giga".
+    """
```
* Add `.. automethod` for methods.
* Document `num_values`.

matthiasdiener replied:
Done

inducer replied:
Doesn't look done. Forgot to push?

inducer commented on mirgecom/profiling.py:
```diff
@@ -51,15 +52,68 @@ class SingleCallKernelProfile:
 
 
+class StatisticsAccumulator:
+    """Class that provides statistical functions for profile results.
```
No real reason to me to make this special-purpose. Move to `mirgecom.tools`?

matthiasdiener replied:
Moved to `mirgecom.utils` (and pushed).

inducer commented on mirgecom/profiling.py:
```diff
@@ -51,15 +52,68 @@ class SingleCallKernelProfile:
+class StatisticsAccumulator:
+    """Class that provides statistical functions for profile results.
+
+    Values are converted to "Giga".
```
Why not introduce a `scale_factor` constructor parameter (defaulting to 1) so as to not wreck the general-purpose-ness?

matthiasdiener replied:
Done.

inducer replied:
Doesn't look done. Forgot to push?

inducer reviewed (commented):
Super close. Just a few tweaks to make `StatisticsAccumulator` general purpose.

matthiasdiener commented:
Should be ready for another review @inducer.
